internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Error("failed to send ctrl-break to process in job")
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again. If there is any left then terminate the job.
+	newPids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.terminate()
+	}
+	return nil
+}
+
+// Returns all of the process IDs in the job object.
+func (job *jobObject) pids() ([]uint32, error) {
+	info := winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST{}
+	var retLen uint32
+	err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		&retLen,
+	)
+
+	// If there are no processes in the job the next call to QueryInformation will just
+	// hang until a memory alloc error. Return here.
+	if err == nil {
+		// Return empty slice instead of nil to play well with the caller of this.
+		// I don't think we should return an error on there being no PIDs.
+		var pids []uint32
+		return pids, nil
+	}
+
+	if err != nil && err != winapi.ERROR_MORE_DATA {
+		return nil, fmt.Errorf("failed initial query for PIDs in job object: %s", err)
+	}
+
+	buf := make([]uintptr, unsafe.Sizeof(info)+unsafe.Sizeof(info.ProcessIdList[0])*uintptr(info.NumberOfAssignedProcesses-1))
+	err = winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&buf[0])),
+		uint32(len(buf)),
+		&retLen)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to query for PIDs in job object: %s", err)
+	}
+
+	bufInfo := (*winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST)(unsafe.Pointer(&buf[0]))
+	rawPids := make([]uintptr, bufInfo.NumberOfProcessIdsInList)
+
+	err = winapi.RtlMoveMemory(
+		(*byte)(unsafe.Pointer(&rawPids[0])),
+		(*byte)(unsafe.Pointer(&bufInfo.ProcessIdList[0])),
+		uintptr(bufInfo.NumberOfProcessIdsInList)*unsafe.Sizeof(rawPids[0]))
+	if err != nil {
+		return nil, fmt.Errorf("failed to move PID info to new buffer: %s", err)
+	}
+
+	pids := make([]uint32, bufInfo.NumberOfProcessIdsInList)
+	for i, rawPid := range rawPids {
+		pids[i] = uint32(rawPid)
+	}
+	return pids, nil
+}
+
+// Polls the IO completion port for notifications. Used for detecting when all of the
+// processes in a job have exited and for (TODO: dcantah) limit thresholds being reached.
+func (job *jobObject) pollIOCP() (uint32, error) {
+	var (
+		overlapped uintptr
+		qty        uint32
+		key        uint32
+	)
+	if job.iocpHandle != 0 {
+		if err := windows.GetQueuedCompletionStatus(job.iocpHandle, &qty, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE); err != nil {
```

Comment:
I'm concerned about this line as well. I don't think it's safe to use unsafe.Pointer here. That needs to be done *in* the syscall call itself, so I think you should actually make an Overlapped struct for this.

Reply:
Okay, this brings up concerns now with the values we pass for the other winapi calls as well.... ah